KAFKA-5325: Avoid and handle exceptions for Kerberos re-login

rajinisivaram opened the pull request:
If producer creates a connection during Kerberos re-login (after logout, before login), there are no principals in the subject and `SaslClientAuthenticator.configure` may throw an exception while trying to determine the principal. A socket channel is created and its key registered with the selector, but the `RuntimeException` thrown leaves the key registered with the selector without adding the channel to the channel list. This results in an infinite loop of `NullPointerExceptions`. The PR applies two fixes:
1. Convert the `RuntimeException` to a meaningful `KafkaException`
2. Handle any exception in `buildChannel`, cleanup and throw `IOException`. Retries will take care of re-connections.

rajinisivaram committed:
KAFKA-5325: Avoid and handle exceptions for Kerberos re-login

rajinisivaram commented:
@ijuma @junrao Can you take a look, please? Thank you...
I couldn't find any easy way to unit test, particularly because we don't have any mock tests for Kerberos code.

ijuma commented on clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java:
```diff
@@ -171,7 +171,7 @@ public void testMissingUsernameSaslPlain() throws Exception {
             selector.connect(node, addr, BUFFER_SIZE, BUFFER_SIZE);
             fail("SASL/PLAIN channel created without username");
-        } catch (KafkaException e) {
+        } catch (IOException e) {
```
Maybe we can assert that we close resources properly?

rajinisivaram replied:
Done.

ijuma commented on clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java:
```diff
@@ -106,13 +107,21 @@ public void configure(TransportLayer transportLayer, PrincipalBuilder principalB
+            // principal (username for PLAIN and SCRAM) is used as authorization id.
+            if (mechanism.equals(SaslConfigs.GSSAPI_MECHANISM)) {
+                try {
+                    Principal clientPrincipal = subject.getPrincipals().iterator().next();
```
It seems a bit odd to invoke `next()` blindly. Can any of these calls throw an exception is `getPrincipals` is not empty?

rajinisivaram replied:
@ijuma If `getPrincipals` is not empty, I dont think the code would throw any exceptions. But `next()` can throw an exception regardless of whether we check `isEmpty()` first since the principals are cleared on a different thread. So rather than  process the failure of `isEmpty()` as well as `next()`, the PR expects `next()` to succeed for `GSSAPI`. If not it indicates that a refresh is in progress and we fail this connection. It is not ideal, but it avoids synchronization (we dont want to block the network thread). It also doesn't make sense to continue with a null `clientPrincipal` in some scenarios for GSSAPI.

ijuma replied:
Thanks for the explanation.

The iterator returned by synchronized connections has to be synchronized by the callers. The code of `SynchronizedCollection` that seems to be used by `getPrincipals`:

```java
public Iterator<E> iterator() {
    return c.iterator(); // Must be manually synched by user!
}
```

So, it looks like we don't really have a choice here.

rajinisivaram replied:
Added synchronization.

ijuma reviewed (commented):
Thanks for the PR, left a couple of minor comments.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java:
```diff
@@ -106,13 +107,21 @@ public void configure(TransportLayer transportLayer, PrincipalBuilder principalB
-            } else {
-                clientPrincipalName = null;
+            // determine client principal from subject for Kerberos. For other mechanisms, the authenticated
+            // principal (username for PLAIN and SCRAM) is used as authorization id.
```
I guess clientPrincipalName is only used for Kerberos and not for other mechanisms like PLAIN or SCRAM? If so, perhaps we could make this clearer in the comment.

rajinisivaram replied:
Updated comment.

junrao reviewed (commented):
@rajinisivaram : Thanks for the patch. LGTM. Just a minor comment. You can merge the patch once it's addressed.

rajinisivaram committed:
KAFKA-5325: Address review comment, fix server-side as well

rajinisivaram commented:
@ijuma @junrao Thank you for the reviews. I have addressed the comments. Also did the same for server-side since I think you could run into the issue on the broker as well.

ijuma reviewed (approved):
LGTM, thanks.

ijuma commented:
Merged to trunk and 0.11.0.

shubhamrander commented:
Hi @rajinisivaram, does it mean that the producer lost connection to kafka servers and is trying to connect during the re-login period. If it is the case then why there are no error logs related to producer connection failures? Can you kindly explain if my above assumption is correct ?
